odoo/addons/base/transpiler/transpiler_js.py

```diff
@@ -1,378 +1,393 @@
-        #    /?                 # Optional /
-        #    (?P<module>        # Capture in a group named module
-        #        \w+            # One or more word character
-        #    )
-        #    /                  # /
-        #    ([\w/]*/)?         # A/few word character or /, zero to many times ended by /
-        #    static             # static
-        #    /                  # /
-        #    (?P<type>          # Capture in a group named type
-        #        src|tests      # src or tests
-        #    )
-        #    (?P<url>           # Capture in a group named url
-        #        /[\w/]*        # A word character or /, zero to many times.
-        #    )
-        #    """, url, re.VERBOSE)
-        if match:
-            url = match["url"]
-            if url.endswith(('/index.js', '/index')):
-                url, _ = url.rsplit('/', 1)
-            if match["type"] == "src":
-                return "@%s%s" % (match['module'], url)
-            else:
-                return "@%s/../tests%s" % (match['module'], url)
+"""
+This code is what let us uses ES6 like JS in odoo.
+Odoo expects all javascript to be divided into functions as "odoo.define(<module_name>, {...});".
+If any file has the /** odoo-module */ on top of it, it will get processed by this class.
+It performs several operations to get from ES6 syntax to the usual odoo one with minimal changes.
+This is done on the fly, this not a pre-processing tool.
+
+Caveat: This is done without a full parser, only using regex. One can only expect to cover as much edge cases
+as possible with reasonable limitations. Also, this only changes imports and exports, so all JS features used in
+the original source need to be supported by the browsers.
+"""
+
+
+def convert_odoo_module(url, content):
+    """
+    Transpile the code through a bunch of operations.
+    The order of the operations does sometimes matter.
+
+    :param content: The original source code
+    :param url: The url of the file in the project
+    :return: The transpiled source code
+    """
+    module_path = url_to_module_path(url)
+    legacy_odoo_define = get_aliased_odoo_define_content(module_path, content)
+
+    new_content = convert_legacy_default_import(content)
+    new_content = convert_basic_import(new_content)
+    new_content = convert_default_import(new_content)
+    new_content = convert_star_import(new_content)
+    new_content = convert_unnamed_relative_import(new_content)
+    new_content = convert_from_export(new_content)
+    new_content = convert_star_from_export(new_content)
+    new_content = convert_relative_require(url, new_content)
+    new_content = remove_index(new_content)
+    new_content = convert_function_and_class_export(new_content)
+    new_content = convert_variable_export(new_content)
+    new_content = convert_list_export(new_content)
+    new_content = convert_default(new_content)
+    new_content = wrap_with_odoo_define(module_path, new_content)
```

Comment:
FWIW here a possibly nice option (I'll let you decide) is to define a list of "transformation passes" (a list of functions), and then use a loop or `reduce` to apply those passes to the content e.g.

```python
passes = [ convert_legacy_default_import, convert_basic_import, ..., partial(wrap_with_odoo_define, module_path)]

for p in passes:
    content = p(content)
```